Use circleCI API to kick off multiple pipelines for different plugins.

jieru-hu opened the pull request:
<!-- Thank you for sending a PR and taking the time to improve Hydra -->
*update 09/17*
CircleCI support helped me find a way to kick off pipelines for forked PRs (after a few back and forth emails.), yayy.

### Test plan:
##### on forked PR
This PR is a forked PR. Pls check the checks sections for all the jobs that was kicked off. one example here, notice the plugin test was kicked off by beloved mecha-hydra.
![image](https://user-images.githubusercontent.com/4642412/93529500-d9509d80-f8f0-11ea-9cf0-3312df6f49bd.png)

##### IN Hydra repo
I created a PR from within Hydra, the checks can be seen in this PR.
https://github.com/facebookresearch/hydra/pull/985


------
*update 09/16*
 - [x] move github token to github secrets https://github.com/facebookresearch/hydra/issues/973
- [x] use deploy key instead of user key in hydra
      I've removed the user key and created a read only deploy key for Hydra cirleCI and kicked off a build here https://app.circleci.com/pipelines/github/facebookresearch/hydra/3484/workflows/d822ee36-a0bb-4a04-9ccd-73a42c2b4db8 (this build runs all the workflows, so if it succeeds it means we are good.)
- [x] some related config needed for forked PR. 

Then we should be good to go with this PR.



*update 09/01*

We found a good solution for running tests in forked PR. 
I'm going to 
- [x] address omry's comments.
- [x] reach out to oss with our plan - done, task in progress
- [x]  submit feature request to circleCI on only sharing selected secrets - done, support ticket created.



---


## TODO 

- [x] Create github bot user and set it up in Hydra (blocked by github support, our bot account was flagged as spam.)



## Motivation

Try out parameterized the nox session.

reduced the CI time from [46m](https://app.circleci.com/pipelines/github/jieru-hu/hydra/143/workflows/e9424f92-6862-4324-b3f2-3513b6fd5ca9) to 16m (disclaimer: one datapoint only)

### Have you read the [Contributing Guidelines on pull requests](https://github.com/facebookresearch/hydra/blob/master/CONTRIBUTING.md)?

Yes/No

## Test Plan

validate yaml locally, and CI


## Related Issues and PRs

(Is this PR part of a group of changes? Link the other relevant PRs and Issues here. Use https://help.github.com/en/articles/closing-issues-using-keywords for help on GitHub syntax)

jieru-hu commented:
One of the things Ive noticed is when there's a lot of [jobs](https://app.circleci.com/pipelines/github/jieru-hu/hydra/144/workflows/f5084e1b-f83d-4353-9b3c-9821c23c6ad9). github shows ```All Checks Passed``` while circleCI is still running some of the jobs.

omry commented on .circleci/config.yml:
```diff
@@ -183,17 +183,17 @@ workflows:
             parameters:
               py_version: ["3.6", "3.7", "3.8"]
-              nox_session: ["lint test_core", "lint_plugins test_plugins", "test_jupyter_notebooks"]
+              nox_session: ["lint", "lint_plugins", "test_core", "test_plugins", "test_jupyter_notebooks"]
```
so actually these are all quick, I think the heavy one is test_plugins. this should be the prize.
It's probably not worth splitting the rest into so many small chunks.

omry commented:
for  each OS and Python verson:

run fast tests + core
run tests for plugin A
run tests for plugin B
run tests for plugin C
...

omry commented:
You are still testing all the plugins at the same time.

jieru-hu commented:
> You are still testing all the plugins at the same time.

The plugins combined are only taking [10 mins](https://app.circleci.com/pipelines/github/jieru-hu/hydra/156/workflows/42be7de4-6222-4a1d-96a9-7ce5c532526b/jobs/1847) though. I'm not sure how much benefit that will bring us to run them all separately.

omry commented:
The plugins will scale up and up as we add more plugins.

omry commented:
Also, for mac they take 30 minutes.
https://app.circleci.com/pipelines/github/jieru-hu/hydra/156/workflows/42be7de4-6222-4a1d-96a9-7ce5c532526b/jobs/1851

omry commented:
You also conveniently linked to a Python 3.6 run which excludes many plugins :)

jieru-hu commented:
> You also conveniently linked to a Python 3.6 run which excludes many plugins :)

Ooops, i promise i didnt do that on purpose :P 

Ok yea, i will work on splitting them

omry commented:
> > You also conveniently linked to a Python 3.6 run which excludes many plugins :)
> 
> Ooops, i promise i didnt do that on purpose :P
> 
> Ok yea, i will work on splitting them

I know, I was joking by hinting that you did :)

jieru-hu commented:
> This makes the noxfile much less generic.
> requesting changes per out discussion. if you are not sure how to solve it let me know.

Update: I'm kinda blocked now since I don't have a circle token to play around the API. I've reached out to our CircleCI sys admin for help.

omry commented:
Okay. this is not the blocker to the Ray plugin. In the interest of time I would rather review this once it's fully implemented.

jieru-hu commented:
update: 

I was able to kick off a pipeline using the circleCI API after being added to the facebookreasearch org. We have a internal ticket open for getting a github bot user for automation purpose.

The request looks like this
```python3
r = requests.post("https://circleci.com/api/v2/project/gh/facebookresearch/hydra/pipeline", headers={"Circle-Token": auth}, data={"branch": "master"})
```

One issue is we need to get the project slug (`facebookresearch/hydra/` in the above example) dynamically, since people mostly like will submit a PR from their own fork of the repo.

I'm following up with circleCI support to see if this is possible.

jieru-hu commented:
per offline chats, a couple of things we need to follow up with circleCI:
1. can we view all the piepliens kicked of by the same PR in a merged view
2.  how do we know how long the tests in total with multiple pipeline runing.

jieru-hu commented:
Overall stats:
---------------------------------
Overall time from [48 m](https://app.circleci.com/pipelines/github/facebookresearch/hydra/3047/workflows/c036ef22-071b-4b0e-9b22-201cc1d73ec5) to 21 m

Overall credit change
From 9744 to 33803


In more details:
---------------------------------

```python3
    r = requests.get(
        f"https://circleci.com/api/v2/insights/gh/facebookresearch/hydra/workflows/build?branch=master",
        headers = {"Circle-Token": auth, "Content-type": "application/json"},
       )
```
Latest build on master
{'id': '0db486e3-bce6-4be0-a337-d5fe6551e84c', 'status': 'success', 'duration': 2823, 'created_at': '2020-08-05T23:07:34.771Z', 'stopped_at': '2020-08-05T23:54:38.614Z', 'credits_used': 9744},



All the workflows from the latest PR:
```python3
    r = requests.get(
        f"https://circleci.com/api/v2/insights/gh/jieru-hu/hydra/workflows/plugin_tests?branch=circleCI",
        headers = {"Circle-Token": auth, "Content-type": "application/json"},
    )
```

```
{'id': '8da223f5-c8c3-4fa1-8632-22a20c83b521', 'status': 'success', 'duration': 1117, 'created_at': '2020-08-07T00:18:53.770Z', 'stopped_at': '2020-08-07T00:37:31.314Z', 'credits_used': 4083}

{'id': 'ab2df82f-9302-4bf8-8d9b-d7efeb71b171', 'status': 'success', 'duration': 1300, 'created_at': '2020-08-07T00:19:32.727Z', 'stopped_at': '2020-08-07T00:41:13.356Z', 'credits_used': 4347}, 

{'id': '583f2c93-596f-49cd-9067-c4e1f324080a', 'status': 'success', 'duration': 1057, 'created_at': '2020-08-07T00:19:33.030Z', 'stopped_at': '2020-08-07T00:37:10.609Z', 'credits_used': 4055}, 


{'id': '19e43c84-4bd1-45c5-b80d-5e6b9d37cd65', 'status': 'success', 'duration': 1074, 'created_at': '2020-08-07T00:19:33.349Z', 'stopped_at': '2020-08-07T00:37:28.328Z', 'credits_used': 4079}, 

{'id': '58a70433-1232-43c5-b76d-6e9843eabc3f', 'status': 'success', 'duration': 1148, 'created_at': '2020-08-07T00:19:33.766Z', 'stopped_at': '2020-08-07T00:38:42.302Z', 'credits_used': 4397}, 

{'id': '438fd149-ceec-4bd4-90ab-caeb89281a15', 'status': 'success', 'duration': 1257, 'created_at': '2020-08-07T00:19:34.035Z', 'stopped_at': '2020-08-07T00:40:31.338Z', 'credits_used': 4094}, 

{'id': '3c1f52f7-19db-447c-918f-90837603399f', 'status': 'success', 'duration': 1097, 'created_at': '2020-08-07T00:19:34.367Z', 'stopped_at': '2020-08-07T00:37:52.295Z', 'credits_used': 4134}, 

{'id': '5ddbdde4-9a06-482d-91cb-f2b95ad1ec1c', 'status': 'success', 'duration': 1288, 'created_at': '2020-08-07T00:19:34.657Z', 'stopped_at': '2020-08-07T00:41:02.742Z', 'credits_used': 4632}, 

```

omry commented:
> Since we have a custom performance plan, I'm not sure how much impact the credit increase has.

I am not sure either, but 4x more credits could be significant. I think this is something we can improve by using caching.
Can you verify with CircleCI support about the implications of such an increase?

jieru-hu commented:
> > Since we have a custom performance plan, I'm not sure how much impact the credit increase has.
> 
> I am not sure either, but 4x more credits could be significant. I think this is something we can improve by using caching.

Yea, I will look into caching.
> Can you verify with CircleCI support about the implications of such an increase?
I did and they suggested I reach out to sys admin. I've made a post on this.

jieru-hu commented:
Update:

I added caching to the MAC os tests and now the runtime has been reduced to 15m! (yayy)

Credit used: 25947


details: 

{'id': '700a99c3-8fb9-47b3-bd7a-820c5d6eefca', 'status': 'success', 'duration': 916, 'created_at': '2020-08-12T00:35:30.000Z', 'stopped_at': '2020-08-12T00:50:46.316Z', 'credits_used': 3757},

{'id': '30c18d9c-c735-4c63-9c77-c73cf1a5f292', 'status': 'success', 'duration': 846, 'created_at': '2020-08-12T00:35:30.247Z', 'stopped_at': '2020-08-12T00:49:36.317Z', 'credits_used': 3465},

{'id': '890bc9dd-2e9b-4c5c-82b7-c9f5cf466b8c', 'status': 'success', 'duration': 832, 'created_at': '2020-08-12T00:35:30.532Z', 'stopped_at': '2020-08-12T00:49:23.063Z', 'credits_used': 3637}, 

{'id': '536ddbc9-0f40-462b-b315-fbd379116c63', 'status': 'success', 'duration': 950, 'created_at': '2020-08-12T00:35:30.862Z', 'stopped_at': '2020-08-12T00:51:21.550Z', 'credits_used': 3851}, 

{'id': '4e12a144-34d2-44e4-9854-a25ac408eb62', 'status': 'success', 'duration': 863, 'created_at': '2020-08-12T00:35:31.152Z', 'stopped_at': '2020-08-12T00:49:54.681Z', 'credits_used': 3634}, 

{'id': 'f91f5a2b-4375-4927-9925-7c4aa5d51fdd', 'status': 'success', 'duration': 851, 'created_at': '2020-08-12T00:35:31.487Z', 'stopped_at': '2020-08-12T00:49:43.426Z', 'credits_used': 3518}, 

{'id': 'd9016129-2264-4d63-83cb-fb2f502b2046', 'status': 'success', 'duration': 1049, 'created_at': '2020-08-12T00:35:31.689Z', 'stopped_at': '2020-08-12T00:53:01.590Z', 'credits_used': 4085},

jieru-hu commented:
per offline chats, next step is to chunk them with a random shuffle.

jieru-hu commented:
Update:

Chunked them into 3 plugins each pipeline, time spent 16m (still a big improvement from 48m :D )
Credit 
```
job: 03e3841e-2f6c-45b7-b1fe-331bc84bbf22, credit: 3963
job: 666399df-57da-4431-a77f-a703c3f1bee6, credit: 4493
job: 4baeec4a-9b7e-493f-b647-983914a0d656, credit: 4072
job: 05e6e54d-33b0-4eb6-b282-e5974e1c0afc, credit: 3695
Total Credit usage: 16223
```

omry commented:
the examples are super fast to run. can probably all be chunked into the same run (and even just added to the primary Hydra run).
in fact I want to move the examples from plugins/examples into examples/plugins.

jieru-hu commented:
> the examples are super fast to run. can probably all be chunked into the same run (and even just added to the primary Hydra run).
> in fact I want to move the examples from plugins/examples into examples/plugins.

All the examples are chunked into one plugin run already.

Time spent: [examples](https://app.circleci.com/pipelines/github/jieru-hu/hydra/350/workflows/03e3841e-2f6c-45b7-b1fe-331bc84bbf22/jobs/4088): 7m58s vs [hydra_nevergrad_sweeper,hydra_ax_sweeper,hydra_submitit_launche](https://app.circleci.com/pipelines/github/jieru-hu/hydra/349/workflows/666399df-57da-4431-a77f-a703c3f1bee6/jobs/4089)r 8m22

omry commented:
Okay.
I added a feature to nox that allow printing the timestamps in the log output.
this can help here, unfortunately it's not released yet. (waiting on https://github.com/theacodes/nox/issues/335).
In the mean time, can you install nox from a specific revision on master so we get it?
you can then change the nox command to add timestamps in circle (--add-timestamps if I remember correctly, see --help if not).

This will help us understand where the rest of the time is going.
my guess is most of it is in the Hydra core tests.

We can speed those with pytest-xdist, see my this #863.

jieru-hu commented:
> Okay.
> I added a feature to nox that allow printing the timestamps in the log output.
> this can help here, unfortunately it's not released yet. (waiting on [theacodes/nox#335](https://github.com/theacodes/nox/issues/335)).
> In the mean time, can you install nox from a specific revision on master so we get it?
> you can then change the nox command to add timestamps in circle (--add-timestamps if I remember correctly, see --help if not).
> 
> This will help us understand where the rest of the time is going.
> my guess is most of it is in the Hydra core tests.
> 
> We can speed those with pytest-xdist, see my this #863.

This is an awesome feature, thx for adding this!

I've added nox installation change to the circleCI and requirments.txt.

Here is one example run of running `nox -s lint test_core test_jupyter_notebooks -ts
` which includes hydra core tests https://app.circleci.com/pipelines/github/jieru-hu/hydra/355/workflows/09f0ce54-d8f5-4f74-8b05-d6c0aa9c0fb3/jobs/4134

times taken for each session:

lint:  27 s (14:30:06 - 14:30:33)
test_core:  3m19s (14:33:53 - 14:37:12)
test_jupyter_notebooks: 39s (14:37:12-14:37:51)

jieru-hu commented:
update:

per offline chats, only install in "pip install" (got rid of "pip isntall -e")

@omry suggest we address pytest-xdist in another PR?

### group 3 plugins together
 time reduced to  **11m53s**
credit: 10845
```
job: 75b48ca0-ddf9-4d4e-a09d-bab30e58dc80, credit: 2423
job: 85643739-6b03-4297-ada6-570423bdbd6f, credit: 2615
job: f53ba5a0-1b60-495a-81c5-060f48c5b3b8, credit: 3114
job: 1b64ae45-57b9-4b1c-9b88-4c90d60d9a56, credit: 2693
```

### group 4 plugins together
 time reduced to  **12m 19s** aaanndd credit **lower** than current runs
credit 
```
job: 841a1787-a7af-4eae-84e3-7b77fefd1a35, credit: 2863
job: 64c0e115-bb08-4de1-be4b-fc0972514375, credit: 3028
job: e2e052f5-1aba-4b39-82fb-fe42b722a50b, credit: 2496
Total Credit usage: 8387
```

omry commented:
Awesome!
A few things:
1. I don't think it's worth caching the checkout dir at this point.
2. you are not caching on Linux, why?

jieru-hu commented:
> Awesome!
> A few things:
> 
> 1. I don't think it's worth caching the checkout dir at this point.
> 2. you are not caching on Linux, why?

On Linux prepping the env takes 30s-ish while restoring cache takes around 30s as well, so I feel there's probably not much there to cache the env.

omry commented:
> > Awesome!
> > A few things:
> > 
> > 1. I don't think it's worth caching the checkout dir at this point.
> > 2. you are not caching on Linux, why?
> 
> On Linux prepping the env takes 30s-ish while restoring cache takes around 30s as well, so I feel there's probably not much there to cache the env.

gotcha.
to reduce clutter in the jobs screen can you not cache the checkout for now? we can add it later if it becomes slow.

omry commented on .circleci/config.yml:
```diff
@@ -8,27 +8,34 @@ commands:
             ~/miniconda3/bin/conda init bash
       - run:
-          name: "Preparing environment"
+          name: "Brew"
```
Can we standardize on:

Resotring environment cache
Preparing environment   (can contain anything we need to do to prepare the env, installing conda, nox, apt-getting etc).
Saving environment cache

omry commented on .circleci/config.yml:
```diff
@@ -8,27 +8,34 @@ commands:
           command: |
-            curl -o Miniconda3-latest-MacOSX-x86_64.sh https://repo.anaconda.com/miniconda/Miniconda3-latest-MacOSX-x86_64.sh
-            bash ./Miniconda3-latest-MacOSX-x86_64.sh -b
+            if [[ -f ~/miniconda3/LICENSE.txt ]] ; then
+              echo "miniconda installed already."
+            else
+              curl -o Miniconda3-latest-MacOSX-x86_64.sh https://repo.anaconda.com/miniconda/Miniconda3-latest-MacOSX-x86_64.sh
+              bash ./Miniconda3-latest-MacOSX-x86_64.sh -b
+            fi
```
You will need to switch to a versioned link for miniconda and use it in the cache tag.
right now, if the miniconda link updates we will not upgrade until the cache times out and suddenly we will start using the new one out of the blue.

jieru-hu replied:
That's a great point! I will update this.

jieru-hu replied:
I was not able to find a way to ping conda version in windows. For now I updated the MAC OS installation.

omry commented on .circleci/config.yml:
```diff
@@ -69,60 +72,155 @@ commands:
             conda create -n hydra python=<< parameters.py_version >> pywin32 -qy
             conda activate hydra
-            pip install nox dataclasses
+            pip install nox@git+git://github.com/theacodes/nox.git@8dcefbf8f59293bc1925dd13376dd6d3aa66e53a
+            pip install dataclasses
```
why do we need to install dataclasses here?

jieru-hu replied:
this was copy pasted from previous versions of the file. `dataclasses` is needed for python 3.6 to work.

omry replied:
needed for what? Doesn't pip installing hydra bring it if it's needed (aka on 3.6)?
try to remove it and see if it works.

jieru-hu replied:
Yea i tried to remove it, but that caused the tests to fail.

omry replied:
curious to know why.
Hydra is installing it's own dependencies inside nox.

omry commented on .circleci/config.yml:
```diff
@@ -69,60 +72,155 @@ commands:
             conda create -n hydra python=<< parameters.py_version >> pywin32 -qy
             conda activate hydra
-            pip install nox dataclasses
+            pip install nox@git+git://github.com/theacodes/nox.git@8dcefbf8f59293bc1925dd13376dd6d3aa66e53a
```
The cache tag should be a function of the things you are installing.
If we change the nox version here it will not get picked up until the cache expires.

jieru-hu replied:
That's a good point! I'm not sure how to do this in a more clean way. For now I defined a parameter for cachekey. Everytime we update any dependency, we will need to bump the cachekey.

omry replied:
The nox version with --add-timestamp was released, you can install nox normally.

omry commented on .circleci/config.yml:
```diff
@@ -69,60 +72,155 @@ commands:
+  plugin_test:
+    type: boolean
+    default: False
+  plugin:
+    type: string
+    default: ""
```
should this be called plugins?

jieru-hu commented:
update:

Since we don't want to share our env variables with forked repo, I suggest:

### for forked PR:
1. for forked PRs we run test the slow way, only one pipeline, skip tests that requires secrets (AWS)
2. Once forked PRs are checked, an Hydra team member could kick off a full run before merging by creating a branch inside Hydra with the forked PR changes, ( we could do something like this https://circleci.com/blog/triggering-trusted-ci-jobs-on-untrusted-forks/). These tests will be run the faster and cheaper way which requires secrets.

### for PR from a Hydra branch.
Tests will be run in this PR's way.

jieru-hu commented:
@omry  thoughts on my last comment?

omry commented:
The comment you made on Friday morning?
nothing yet, I was busy on Friday and am not working today.

jieru-hu commented:
@omry 

I've updated the test plan and latest status in description box on the top of PR. :)

omry commented on .circleci/config.yml:
```diff
@@ -59,71 +71,161 @@ commands:
+          key: -<< pipeline.parameters.cache_key_version >>-win-sys-{{ .Branch }}-<< parameters.py_version >>
       - run:
-          name: Preparing system
+          name: "Preparing environment - system"
```
You don't normally need to quote strings in YAML. this can be unquoted.

jieru-hu replied:
good to know. Removed all of them.

omry commented on .circleci/config.yml:
```diff
@@ -59,71 +71,161 @@ commands:
             choco install -y --no-progress openssl javaruntime
       - run:
-          name: Preparing conda environment
+          name: "Preparing environment - Hydra"
```
ditto

jieru-hu replied:
👍

omry commented on .circleci/config.yml:
```diff
@@ -59,71 +71,161 @@ commands:
+          py_version: << parameters.py_version >>
+      - run:
+          name: Testing Hydra
+          no_output_timeout: 20m
```
inconsistent timeout. on purpose?

jieru-hu replied:
Yea, this was the original value in the config file, I just copy pasted them :)

omry replied:
but you did change timeouts in other locations in the circleci file.

jieru-hu replied:
Could you point me to that?

I added extra jobs for test plugins, but in general I just put 10m for macos and 20m for windows to be consistent with the previous version.

omry replied:
The no-output-timeouts of the previous version are irrelevant. you improved the speed a lot.
Make it 10m everywhere.

omry commented on .circleci/config.yml:
```diff
@@ -151,5 +255,26 @@ workflows:
+          matrix:
+            parameters:
+              py_version: ["3.6", "3.7", "3.8"]
+              test_plugin: [<< pipeline.parameters.plugin >>]
```
should this be test_plugins?

jieru-hu replied:
do you mean re-naming this param?

omry replied:
yes. this is not a plugin right? it's a list of plugins.

jieru-hu replied:
good point. fixed.

omry commented on noxfile.py:
```diff
@@ -331,8 +336,8 @@ def _get_standalone_apps_dirs():
     "install_cmd",
-    PLUGINS_INSTALL_COMMANDS,
-    ids=[" ".join(x) for x in PLUGINS_INSTALL_COMMANDS],
+    (INSTALL_COMMAND,),
```
you can clean this up. this is no longer a list. (you are making it a list lf 1.
you can just take it out of the parameterization.

jieru-hu replied:
sure, will do!

jieru-hu replied:
cleaned up

omry commented on requirements/dev.txt:
```diff
@@ -6,7 +6,7 @@ flake8
 isort@git+git://github.com/timothycrosley/isort.git@c54b3dd#egg=isort
 mypy
-nox
+nox@git+git://github.com/theacodes/nox.git@8dcefbf8f59293bc1925dd13376dd6d3aa66e53a
```
nox ox x

jieru-hu replied:
? not sure i understand this comment :P

omry replied:
install regular nox.

jieru-hu replied:
got it

omry commented on tools/ci/circleci_pipeline.py:
```diff
@@ -0,0 +1,69 @@
+    pr_number = os.environ.get("CIRCLE_PR_NUMBER", "")
+
+    # https://support.circleci.com/hc/en-us/articles/360049841151
+    if pr_number:
+        branch += "/head"
```
holy hacks :)

jieru-hu replied:
whatever it takes :)

omry commented on .circleci/config.yml:
```diff
@@ -1,129 +1,231 @@
             conda activate hydra
-            nox
-            exit $LASTEXITCODE
+            pip install nox@git+git://github.com/theacodes/nox.git@8dcefbf8f59293bc1925dd13376dd6d3aa66e53a
```
```suggestion
            pip install nox
```

omry commented on .circleci/config.yml:
```diff
@@ -1,129 +1,231 @@
+          name: Testing Hydra
+          command: |
+            export NOX_PYTHON_VERSIONS=<< parameters.py_version >>
+            pip install nox@git+git://github.com/theacodes/nox.git@8dcefbf8f59293bc1925dd13376dd6d3aa66e53a
```
```suggestion
            pip install nox
```

omry commented on .circleci/config.yml:
```diff
@@ -1,129 +1,231 @@
+              export NOX_PYTHON_VERSIONS=<< parameters.py_version >>
+              export PLUGINS=<< parameters.test_plugin >>
+              conda activate hydra
+              pip install nox@git+git://github.com/theacodes/nox.git@8dcefbf8f59293bc1925dd13376dd6d3aa66e53a
```
```suggestion
              pip install nox
```

omry commented on .circleci/config.yml:
```diff
@@ -1,129 +1,231 @@
+          command: |
+                export NOX_PYTHON_VERSIONS=<< parameters.py_version >>
+                export PLUGINS=<< parameters.test_plugin >>
+                pip install nox@git+git://github.com/theacodes/nox.git@8dcefbf8f59293bc1925dd13376dd6d3aa66e53a
```
```suggestion
                pip install nox
```

omry commented on noxfile.py:
```diff
@@ -329,14 +334,9 @@ def _get_standalone_apps_dirs():
 
 
 @nox.session(python=PYTHON_VERSIONS)
-@nox.parametrize(
-    "install_cmd",
-    PLUGINS_INSTALL_COMMANDS,
-    ids=[" ".join(x) for x in PLUGINS_INSTALL_COMMANDS],
-)
-def test_core(session, install_cmd):
```
I bet this makes the session names much nicer!

jieru-hu commented:
re: dataclass dependency

Somehow github no longer shows that thread on why we pip install dataclasses

Yea, Im not sure why either it doesnt work for python 3.6. I will file a issue follow up separately on that. 

The command to install dataclasses has been there before all the config file changes https://github.com/facebookresearch/hydra/blob/6b960fb0ff4306ff7f76d8a9c500b19d71b2c75b/.circleci/config.yml#L29
It's not something new this diff introduced.

shagunsodhani commented on .circleci/config.yml:
```diff
@@ -1,129 +1,231 @@
+    default: ""
+  cache_key_version:
+    type: string
+    default: "v0"
```
quotes are generally optional in yaml files.

jieru-hu replied:
thx, i missed this one :)

omry reviewed (approved):
YaY!

omry commented:
One additional low hanging fruit for later is the test coverage.
we are currently testing twice, once with coverage enabled and once without. I have yet to see a case where the coverage caused issues so we can just run it always.
We should also start surfacing the coverage results, but that's a different problem.

jieru-hu commented:
> we are currently testing twice, once with coverage enabled and once without. I have yet to see a case where the coverage caused issues so we can just run it always.

sounds good. I will create a issue to track this :)
